File: historyscan-back/src/main/java/io/sch/historyscan/domain/contexts/analysis/clocrevisions/filesystem/CodeBaseFile.java
package io.sch.historyscan.domain.contexts.analysis.clocrevisions.filesystem;

import java.io.File;
import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import static io.sch.historyscan.domain.contexts.analysis.common.EnumIgnoredCodeBaseFiles.isIgnored;

public class CodeBaseFile {
    private final File rootFile;
    private final RootFolder rootFolder;
    private final List<FileInfo> ignoredFiles;
    private final String codebasesPath;

    public CodeBaseFile(File rootFile, RootFolder rootFolder, String codebasesPath) {
        this.rootFile = rootFile;
        this.rootFolder = rootFolder;
        this.codebasesPath = codebasesPath;
        this.ignoredFiles = new ArrayList<>();
    }

    public boolean hasSameNameAsCodeBase() {
        return rootFile.getName().equals(rootFolder.getCodebaseName());
    }

    public List<FileInfo> filteredChildren() {
        try (var codeBaseFiles = Files.walk(rootFile.toPath())) {
            return codeBaseFiles.map(Path::toFile)
                    .filter(file -> new FilePath(file, rootFolder.getValue(), rootFolder.getCodebaseName(), codebasesPath)
                            .isValid())
                    .map(this::child)
                    .filter(Optional::isPresent)
                    .map(Optional::get)
                    .toList();
        } catch (IOException e) {
            throw new CodebasePathCanNotBeRead("Error while reading codebase files tree", e);
        }
    }

    private Optional<FileInfo> child(File codeBaseFile) {
        var filePath = new FilePath(codeBaseFile, rootFolder.getValue(), rootFolder.getCodebaseName(), codebasesPath);
        var path = filePath.pathFromCodebaseName(codeBaseFile);
        var currentNbLines = nbLinesOfCode(codeBaseFile);
        final FileInfo fileInfo = new FileInfo(codeBaseFile.getName(), path, codeBaseFile.isFile(), currentNbLines);
        if (this.filterIgnoredFiles(codeBaseFile, path, fileInfo)) {
            return Optional.of(fileInfo);
        }
        return Optional.empty();
    }

    private long nbLinesOfCode(File codeBaseFile) {
        if (codeBaseFile.isDirectory()) {
            return 0;
        }
        try {
            var lines = Files.readAllLines(codeBaseFile.toPath(), Charset.defaultCharset());
            return lines.size();
        } catch (IOException ex) {
            return 0;
        }
    }

    private boolean filterIgnoredFiles(File currentFile, String path, FileInfo fileInfo) {
        var isIgnored = isIgnored(rootFile.getPath(), path, currentFile.isFile());
        if (isIgnored && currentFile.isFile()) {
            this.ignoredFiles.add(fileInfo);
        }
        return !isIgnored;
    }

    public List<FileInfo> getIgnoredFiles() {
        return ignoredFiles;
    }
}

File: historyscan-back/src/main/kotlin/io/sch/historyscan/domain/contexts/analysis/clocrevisions/filesystem/CodeBaseFile.kt
package io.sch.historyscan.domain.contexts.analysis.clocrevisions.filesystem

import io.sch.historyscan.domain.contexts.analysis.common.EnumIgnoredCodeBaseFiles
import java.io.File
import java.io.IOException
import java.nio.charset.Charset
import java.nio.file.Files
import java.nio.file.Path
import java.util.*

class CodeBaseFile(private val rootFile: File, private val rootFolder: RootFolder, private val codebasesPath: String) {
    val ignoredFiles: MutableList<FileInfo> = ArrayList()

    fun hasSameNameAsCodeBase(): Boolean {
        return rootFile.name == rootFolder.codebaseName
    }

    fun filteredChildren(): List<FileInfo> {
        try {
            Files.walk(rootFile.toPath()).use { codeBaseFiles ->
                return codeBaseFiles.map { obj: Path -> obj.toFile() }
                        .filter { file: File ->
                            FilePath(file, rootFolder.value, rootFolder.codebaseName, codebasesPath)
                                    .isValid
                        }
                        .map { codeBaseFile: File -> this.child(codeBaseFile) }
                        .map { it!! }
                        .toList()
            }
        } catch (e: IOException) {
            throw CodebasePathCanNotBeRead("Error while reading codebase files tree", e)
        }
    }

    private fun child(codeBaseFile: File): FileInfo? {
        val filePath = FilePath(codeBaseFile, rootFolder.value, rootFolder.codebaseName, codebasesPath)
        val path = filePath.pathFromCodebaseName(codeBaseFile)
        val currentNbLines = nbLinesOfCode(codeBaseFile)
        val fileInfo = FileInfo(codeBaseFile.name, path, codeBaseFile.isFile, currentNbLines)
        if (this.filterIgnoredFiles(codeBaseFile, path, fileInfo)) {
            return fileInfo
        }
        return null
    }

    private fun nbLinesOfCode(codeBaseFile: File): Long {
        if (codeBaseFile.isDirectory) {
            return 0
        }
        try {
            val lines = Files.readAllLines(codeBaseFile.toPath(), Charset.defaultCharset())
            return lines.size.toLong()
        } catch (ex: IOException) {
            return 0
        }
    }

    private fun filterIgnoredFiles(currentFile: File, path: String, fileInfo: FileInfo): Boolean {
        val isIgnored: Boolean = EnumIgnoredCodeBaseFiles.isIgnored(rootFile.path, path, currentFile.isFile)
        if (isIgnored && currentFile.isFile) {
            ignoredFiles.add(fileInfo)
        }
        return !isIgnored
    }
}
